include/deal.II/base/mpi.templates.h

```diff
@@ -457,80 +457,103 @@ namespace Utilities
             {
-              MPI_Status status;
-              const auto ierr_1 = MPI_Probe(rank_send,
-                                            internal::Tags::compute_union,
-                                            comm,
-                                            &status);
-              AssertThrowMPI(ierr_1);
-
-              int        amount;
-              const auto ierr_2 =
-                MPI_Get_count(&status,
-                              internal::mpi_type_id(result.data()),
-                              &amount);
-              AssertThrowMPI(ierr_2);
-
-              const unsigned int old_size = result.size();
-              result.resize(old_size + amount);
-
-              const auto ierr_3 = MPI_Recv(result.data() + old_size,
-                                           amount,
-                                           internal::mpi_type_id(result.data()),
-                                           status.MPI_SOURCE,
-                                           status.MPI_TAG,
-                                           comm,
-                                           MPI_STATUS_IGNORE);
-              AssertThrowMPI(ierr_3);
-
-              std::sort(result.begin(), result.end());
-              result.erase(std::unique(result.begin(), result.end()),
-                           result.end());
+              const unsigned int rank_recv =
+                (2 * stride) * (rank / (2 * stride));
+              const unsigned int rank_send = rank_recv + stride;
+
+              if (rank_send >= nproc) // nothing to do
+                continue;
+
+              if (rank_recv == rank) // process receives data
+                {
+                  MPI_Status status;
+                  const auto ierr_1 = MPI_Probe(rank_send,
+                                                internal::Tags::compute_union,
+                                                comm,
+                                                &status);
+                  AssertThrowMPI(ierr_1);
+
+                  int        amount;
+                  const auto ierr_2 = MPI_Get_count(&status, MPI_CHAR, &amount);
+                  AssertThrowMPI(ierr_2);
+
+                  std::vector<char> temp(amount);
+
+                  const auto ierr_3 = MPI_Recv(temp.data(),
+                                               amount,
+                                               MPI_CHAR,
+                                               status.MPI_SOURCE,
+                                               status.MPI_TAG,
+                                               comm,
+                                               MPI_STATUS_IGNORE);
+                  AssertThrowMPI(ierr_3);
+
+                  result = binary_op(result, Utilities::unpack<T>(temp, false));
+                }
+              else if (rank_send == rank) // process sends data
+                {
+                  auto temp = Utilities::pack(result, false);
```

Comment:
Please spell out the type
```suggestion
                  const std::vector<char> temp = Utilities::pack(result, false);
```